### PR TITLE
Automaton grammar generation

### PR DESCRIPTION
**Pull request by bebidek**
This is draft pull request because it depends on regex-additions PR. When regex-additions are merged, I will change it.

### CODE REVIEW

File: app/src/main/kotlin/compiler/common/dfa/CompositeDfa.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package compiler.common.dfa
+
+import compiler.common.dfa.state_dfa.Dfa
+import compiler.common.dfa.state_dfa.DfaState
+
+class CompositeDfa<S : Comparable<S>, R, IR> (private val components: List<Pair<Dfa<S, IR>, R>>) : Dfa<S, R> {
```
Comment (ggawryal) [inline app/src/main/kotlin/compiler/common/dfa/CompositeDfa.kt:6]: Currently we ignore the result type for composed dfas - I guess we can change `IR` to `Unit` (suggested) or modify `components` type to `List<Triple<Dfa<S, IR>, IR, R>>` (probably we don't need such functionality)

**Review by ggawryal — CHANGES_REQUESTED** [review]
Great contribution, left 1 comment. Regex additions have been merged, so please merge with current `master`.

File: app/src/main/kotlin/compiler/common/dfa/CompositeDfa.kt (hunk 1 above)
Comment (bebidek) [reply app/src/main/kotlin/compiler/common/dfa/CompositeDfa.kt]: I changed `IR` to `*` (so it's basically your first suggestion but hardcoding `Unit` when it's not necessary seems little weird to me).

**Comment by bebidek** [comment]
I changed `AutomatonGrammar.startState` to `start` because grammar doesn't have states (maybe, if `start` is too vague, it could be `startSymbol`).

**Review by ggawryal — APPROVED** [review]
LGTM

**Pull request merged**